packages/models/src/event/hash.js

```diff
@@ -0,0 +1,40 @@
+import sha256 from 'crypto-js/sha256';
+import { buildQueryAST } from '../util';
+
+// returns a JSON of SQL query AST
+// with all literals replaced by variables
+// and all variable names removed
+function abstractSqlAstJSON(query) {
+  const ast = buildQueryAST(query);
+  return JSON.stringify(ast, (_, value) => {
+    switch (value.type) {
+      case 'variable':
+      case 'literal':
+        return { type: 'variable' };
+      default:
+        return value;
+    }
+  });
+}
+
+// Returns a string suitable for durable identification of a call event
+// that's independent of parameters and return values.
+// For SQL queries, it's a JSON of abstract query AST.
+// For HTTP queries, it's the method plus normalized path info.
+// For function calls it's the qualified function id.
+function callEventToString(event) {
+  const { sqlQuery } = event;
+  if (sqlQuery) return abstractSqlAstJSON(sqlQuery);
+  return event.toString();
```

Comment:
It seems like relying on `event.toString()` might be risky for the purposes of a stable hash. `toString` is probably thought of as more of a debugging / informational method and someone might change `toString` for HTTP events - and break hashing.

Reply:
You might be right. I've been a bit on the fence about it.